### ISSUE TITLE
[Proposal]Enhance the class SparseAutoencoder and SoftmaxRegression 

### ISSUE THREAD

**Issue by stereomatchingkiss** [mlpack/mlpack#454]
As one of the user of mlpack, I would like to have following features

1 : allow users to get the trained parameters of SparseAutoencoder and SoftmaxRegression 

According to UFLDL, if you want to finetune the stacked autoencoder, you will need the trained parameters of SparseAutoencoder and SoftmaxRegression, so I suggest add two functions to get and set the trained parameters of SparseAutoencoder and SoftmaxRegression 

ex : 
arma::mat parameters() const;
void parameters(arma::mat const &parameters);

2 : provide a default constructor for class SparseAutoencoder and SoftmaxRegression
This allow users to train the SparseAutoencoder or SoftmaxRegression later on, so the users do not need to wrap them in smart pointer just to delay the initialization time.

3 : This one is related to suggestion 2, provide a train function for SparseAutoencoder and SoftmaxRegression, this way the users can train the data whenever they like

ex : 
//for SoftMaxRegression

//this one will use a random initialPoint to train
void train(arma::mat const &input, arma::vec const &label, size_t numOfLabels);

//this one can setup the initialPoint
void train(arma::mat const &input, arma::const &initialPoint, 
arma::vec const &label, size_t numOfLabels);

template<typename SparseAutoEncoderFunc>
void train(OptimizerType<SparseAutoEncoderFunc>& optimizer);

template<typename SparseAutoEncoderFunc>
 void train(OptimizerType<SparseAutoEncoderFunc>& optimizer, arma::mat const &initialPoint);

//for SparseAutoencoder

//this one will use a random parameters to train
void train(arma::mat const &data, size_t hiddenSize);

//this one can reuse the trained parameter
void train(const arma::mat &data, const arma::mat &parameters,
               size_t hiddenSize);

The programs should be able to detect the visibleSize from the input(input.n_rows), so the users do not need to set it

4 : Add "m_" before the data member of the class, by now SoftMaxRegression and SparseAutoencoder will use the same name to initialize the data member.

ex : 
SparseAutoencoder(const arma::mat& data,
                                 const size_t visibleSize,
                                 const size_t hiddenSize,
                                 double lambda,
                                 double beta,
                                 double rho) :
    visibleSize(visibleSize),
    hiddenSize(hiddenSize),
    lambda(lambda),
    beta(beta),
    rho(rho)

I think adding "m_" before the data member could make developer easier to identify which one is data member

Exactly, these requirements almost finished on my computer, if you agree with the suggestions(which one?), I will send a pull request.Thank you very much

**Comment by rcurtin** [comment]
It's probably worth taking a look at the semi-finished https://github.com/mlpack/mlpack/wiki/ProposedNewDesignGuidelines and the style guide at https://github.com/mlpack/mlpack/wiki/StyleGuidelines.

> 1 : allow users to get the trained parameters of SparseAutoencoder and SoftmaxRegression
> ex :
> arma::mat parameters() const;
> void parameters(arma::mat const &parameters);

Fine by me, but make sure to pass by reference:

```
arma::mat& Parameters(); // for modification
const arma::mat& Parameters() const; // for access
```
> 2 : provide a default constructor for class SparseAutoencoder and SoftmaxRegression
> This allow users to train the SparseAutoencoder or SoftmaxRegression later on, so the users do not need to wrap them in smart pointer just to delay the initialization time.

We need to be slightly careful with this one.  If we provide a default constructor, what should be returned is a valid sparse autoencoder or a valid softmax regression object.  The reason for this is that we should avoid situations where a user can have an invalid `SparseAutoencoder` or `SoftmaxRegression` object.  What do you think?  I am still thinking about the right way to do default constructors (or how to handle default-constructed mlpack objects).

> 3 : This one is related to suggestion 2, provide a train function for SparseAutoencoder and SoftmaxRegression, this way the users can train the data whenever they like

Yes, this is a good idea.  The constructors will be able to be refactored to call `Train()`, too.  Some comments on your proposed functions below:

> ex :
> //for SoftMaxRegression

> //this one will use a random initialPoint to train
> void train(arma::mat const &input, arma::vec const &label, size_t numOfLabels);
> //this one can setup the initialPoint
> void train(arma::mat const &input, arma::const &initialPoint,
> arma::vec const &label, size_t numOfLabels);

At the point of calling `Train()`, the `SoftmaxRegression` model should already be a valid object.  So setting the initial point for the training shouldn't be necessary; instead, `Train()` should use the current model parameters as the initial point for the training.  As a nice side effect, this means that you can train on multiple datasets sequentially and get a coherent model in the end.

> template
> void train(OptimizerType& optimizer);
> template
> void train(OptimizerType& optimizer, arma::mat const &initialPoint);

For this particular overload, it's worth pointing out that `optimizer` holds an `initialPoint` inside of it, so the initial point overload ends up being redundant here.

I'd also consider adding a single-point training function:

```
template<typename VecType>
void Train(const VecType& point, const size_t label);
```

> //for SparseAutoencoder

> //this one will use a random parameters to train
> void train(arma::mat const &data, size_t hiddenSize);
> //this one can reuse the trained parameter
> void train(const arma::mat &data, const arma::mat &parameters,
> size_t hiddenSize);

See the comments on `SoftmaxRegression::Train()`---at this point, we should already have a coherent `SparseAutoencoder` model, so we should start our training from the current model parameters.

> 4 : Add "m_" before the data member of the class, by now SoftMaxRegression and SparseAutoencoder will use the same name to initialize the data member.

I disagree with this one; the style guidelines already suggest naming conventions, and then access to the internal parameters is simply done by capitalizing the first letter of the parameter.  For instance, if a class holds a member `matrix`, then this can be accessed/modified with the `Matrix()` function.  Also it would be a truly tremendous amount of work to refactor all mlpack code to adhere to a new naming scheme. :)

If you want to send pull requests for pieces at a time, I'll merge them in (as long as they're specific to the existing SparseAutoencoder and SoftmaxRegression code, otherwise it may have to wait on someone else's approval too).

**Comment by stereomatchingkiss** [comment]
Finished the proposed and style guide, some questions here

void(double value) and void(const double value) are the same thing, compiler treat them as same symbol, and since the copy of primitive would not have any side effect, could I just declare them as non const?like void(double value) rather than void(const double value).

About the default constructor, I cannot find a good way to design a default constructor to fulfill the requirements for mlpack yet, so I think it is better to remain the old way(no default constructor)

About the naming convention, I am agree with you, it is better to stick with current style.

The last thing is, if you do not care about breaking the old api, I would like to replace the SparseAutoencder and SparseAutoencoderFunction under the folder methods/sparse_autoencoder

**Comment by stereomatchingkiss** [comment]
I forgot to said, rather than add a default constructor, I would suggest adding one more constructors , this way the users can reuse the trained parameters to train the model if they want to.

For SparseAutoencder

when user do not want to retrain the model but reuse the trained parameters to get the features

SparseAutoencoder<> sae(""smoke_params_of_sae");

Same as SoftmaxRegression

SoftmaxRegression<> sm("smoke_params_of_sm"); 

This way the users can reuse the trained more easier, I intent to use boost::serialization to store the data of the class, what do you think about it?

**Comment by rcurtin** [comment]
> void(double value) and void(const double value) are the same thing, compiler treat them as same symbol, and since the copy of primitive would not have any side effect, could I just declare them as non const?like void(double value) rather than void(const double value).

Yeah, these are the same.  I often write 'const double' just to denote that the double won't be modified, but it doesn't really make a difference either way.

> About the default constructor, I cannot find a good way to design a default constructor to fulfill the requirements for mlpack yet, so I think it is better to remain the old way(no default constructor)

In some cases I've made a default constructor that takes the model's size but nothing else.  Take a look at the constructor I made for `NaiveBayesClassifier` yesterday; maybe that is a strategy that could be applied here?

https://github.com/mlpack/mlpack/commit/4a469b2733c122870274ab813528182157b12211#diff-9ce92a606304271ed37a8531bb3de99dR69

I think maybe instead of default constructors, simple constructors that don't take data can be used.

> The last thing is, if you do not care about breaking the old api, I would like to replace the SparseAutoencder and SparseAutoencoderFunction under the folder methods/sparse_autoencoder

Replace it with what?  Do you mean moving it to `src/mlpack/methods/ann/` or something?  I've got no problem changing the API; we're preparing for a release that will not be reverse-compatible (API-wise) anyway.  Lots of the old design abstractions turned out to not be the best ones.

> I forgot to said, rather than add a default constructor, I would suggest adding one more constructors , this way the users can reuse the trained parameters to train the model if they want to.

I didn't completely understand this idea, could you elaborate more please?

> But this also mean these two classes better offer users "save" and "load" functions.
> sm.save("smoke_params_of_sm");
> sm.load("smoke_params_of_sm");

Why not just use the existing `data::Load()` function that works with `boost::serialization` once `Serialize()` is implemented?  That would look like this:

```
data::Save("model_file.xml" /* or .bin or .txt */, "softmax_regression", sm);
data::Load("model_file.xml", "softmax_regression", sm);
```

> This way the users can reuse the trained results easier, I intent to use boost::serialization to store and save the data of the class, what do you think about it?

`boost::serialization` is pretty nice to use; you can take a look at the other classes I've used for examples (for instance, `NaiveBayesClassifier`, which I just did), or refer to their documentation.  The only difference is the serialization shim that means you can write a `Serialize()` method instead of a `serialize()` method (although, as a technical note, if you write a `serialize()` method, it will still work.  If you like template metaprogramming, take a look at `src/mlpack/core/data/serialization_shim.hpp`; it was a lot of fun to write. :) ).

**Comment by stereomatchingkiss** [comment]
>I didn't completely understand this idea, could you elaborate more please?
Ok, I mean provide a default constructor which takes the name of the files

    SoftmaxRegression(std::string const &fileName);

This way the users can reuse the training results in their codes easier(without training).

    SoftmaxRegression<> sm("smoke_params_of_sm"); 
    sm.Predict(data, result);

>Replace it with what? Do you mean moving it to src/mlpack/methods/ann/ or something?
Replace it with the original files under the folder method/sparse_autoencoder
Since backward compatibility is not an issue, I will place them int the folder "method/sparse_autoencoder"

>I think maybe instead of default constructors, simple constructors that don't take data can be used.
Thanks, will reference from it

>Why not just use the existing data::Load() function that works with boost::serialization once Serialize() is implemented?
Thanks:), I did not noticed this solution

**Comment by rcurtin** [comment]
With #356 merged, I think the SoftmaxRegression part of this issue is fixed and improved.  So I guess that just leaves the sparse autoencoder. :)

**Comment by rcurtin** [comment]
Oops, I meant #457 not #356...

**Comment by rcurtin** [comment]
When #451 was merged this became completed.  (I'm only recording this for later documentation, should anyone come across this.)
